topics/command-tips.md

```diff
@@ -0,0 +1,88 @@
+1. The command has key(s). In this case the command goes to a single shard, determined by the hslot(s) of the key(s)
+2. The command doesn't have key(s). In this case the command goes to an arbitrary shard (usually the one with the lowest hash slot). Different key-less commands should always go to the same shard.
+
+If the proxy/client need to behave differently we must specify an option for `request_policy`:
+- **all_shards** - Forward to all shards (`CONFIG SET`). Usually used on key-less command. The operation is atomic on all shards.
+- **few_shards** - Forward to several shards, used by multi-key commands where each key is handled separately (`MSET`, `MGET`, `DEL`, etc.). i.e. unlike `SUNIONSTORE` which must be sent to one shard.
+- **non_trivial** - Indicates a non-trivial form of request policy. Example: `SCAN`
+
+### reply_policy
+
+This tip should help clients/proxies to know how to aggregate the replies of a command that was sent to multiple shards.
+The default behavior (i.e. if the `request_policy` tip is absent) applies only when the reply is some sort of collection (array, set, map) and it devides into two cases:
+1. The command doesn't have key(s). In this case we append the array replies in random order (e.g. `KEYS`)
+2. The command has key(s).  In this case we append the array replies in the original order of the request's keys (e.g. `MGET`, but not `MSET` and `DEL` which don't return an array)
+
+If the reply is not a collection, or if the proxy/client need to behave differently we must specify an option for `reply_policy`:
+- **one_succeeded** - Return success if at least one shard didn't reply with an error. The client/proxy should reply with the first reply it gets which isn't an error, or with any of the errors, if they all responded with errors. Example: `SCRIPT KILL` (usually the script is loaded to all shards, but runs only on one. `SCRIPT KILL` is sent to all shards)
+- **all_succeeded** - Return success if none of the shards replied with an error. if one replied with an error, return that error (either one of the errors), if they all succeeded, return the successful reply (either one). Examples: `CONFIG SET`, `SCRIPT FLUSH`, `SCRIPT LOAD`
+- **agg_logical_and** - Preform a logical AND on the replies (replies must be numerical, usually just 0/1). Example: `SCRIPT EXISTS`, which returns an array of 0/1 indicating which of the provided scripts exist. The aggregated response will be 1 iff all shards have the script.
+- **agg_logical_or** - Preform a logical OR on the replies (replies must be numerical, usually just 0/1).
+- **agg_min** - Perform a minimum on replies (replies must be numerical). Example: `WAIT` (returns the lowest number among the ones the shards' replies).
+- **agg_max** - Perform a maximum on replies (replies must be numerical).
+- **agg_sum** - Sums the integer values returned by the shards. Example: `DBSIZE`
+- **non_trivial** - Indicates a non-trivial form of reply policy. Example: `INFO`
```

Comment:
Can we assume that if no policy is specified, it's by definition "non-trivial"? Applies also for request policy.

Reply:
we chose the default ones because they are the most common. if non-trivial is the default we will have to mention either FIRST_SHARD or SLOT_SHARD as request_policy in the json files of the vast majority of commands

Reply:
@oranagra WDYT?

Reply:
i agree with Guy.. i rather mention a sane default, and then flag the few command that can't match it and may need special handling.

Reply:
how about "special"?

Reply:
we went with "special"